pkg/controllers/clusterexternalsecret/clusterexternalsecret_controller.go

```diff
@@ -0,0 +1,145 @@
+	errLabelMap           = "unable to get map from labels"
+	errNamespaces         = "could not get namespaces from selector"
+	errGetExistingES      = "could not get existing ExternalSecret"
+	errCreatingOrUpdating = "could not create or update ExternalSecret"
+	errSetCtrlReference   = "could not set the controller owner reference"
+)
+
+//+kubebuilder:rbac:groups=external-secrets.io,resources=clusterexternalsecrets,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=external-secrets.io,resources=clusterexternalsecrets/status,verbs=get;update;patch
+
+func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("ClusterExternalSecret", req.NamespacedName)
+
+	var clusterExternalSecret esv1alpha1.ClusterExternalSecret
+
+	err := r.Get(ctx, req.NamespacedName, &clusterExternalSecret)
+	if apierrors.IsNotFound(err) {
+		return ctrl.Result{}, nil
+	} else if err != nil {
+		log.Error(err, errGetCES)
+		return ctrl.Result{}, nil
+	}
+
+	p := client.MergeFrom(clusterExternalSecret.DeepCopy())
+	defer func() {
+		err = r.Status().Patch(ctx, &clusterExternalSecret, p)
+		if err != nil {
+			log.Error(err, errPatchStatus)
+		}
+	}()
+
+	// Fetch Namespaces to grab ExternalSecrets
+	genClient := kubernetes.NewForConfigOrDie(ctrl.GetConfigOrDie())
+	namespaces := genClient.CoreV1().Namespaces()
+
+	labelMap, err := metav1.LabelSelectorAsMap(&clusterExternalSecret.Spec.NamespaceSelector)
+	if err != nil {
+		log.Error(err, errLabelMap)
+		return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+	}
+
+	namespaceList, err := namespaces.List(ctx, metav1.ListOptions{LabelSelector: labels.SelectorFromSet(labelMap).String()})
+	if err != nil {
+		log.Error(err, errNamespaces)
+		return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+	}
+
+	esName := clusterExternalSecret.Spec.ExternalSecretName
+	if esName == "" {
+		esName = clusterExternalSecret.ObjectMeta.Name
+	}
+
+	for _, namespace := range namespaceList.Items {
+		var existingES esv1alpha1.ExternalSecret
+		err = r.Get(ctx, types.NamespacedName{
+			Name:      esName,
+			Namespace: namespace.Name,
+		}, &existingES)
+
+		if err != nil && !apierrors.IsNotFound(err) {
+			log.Error(err, errGetExistingES)
+		}
+
+		externalSecret := esv1alpha1.ExternalSecret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      esName,
+				Namespace: namespace.Name,
+			},
+			Spec: clusterExternalSecret.Spec.ExternalSecretSpec,
+		}
+
+		err = controllerutil.SetControllerReference(&clusterExternalSecret, &externalSecret, r.Scheme)
+		if err != nil {
+			log.Error(err, errSetCtrlReference)
+			return ctrl.Result{RequeueAfter: r.RequeueInterval}, nil
+		}
+
+		mutateFunc := func() error {
+			externalSecret.Spec = clusterExternalSecret.Spec.ExternalSecretSpec
+			return nil
+		}
+
+		// An empty mutate func as nothing needs to happen currently
+		_, err = ctrl.CreateOrUpdate(ctx, r.Client, &externalSecret, mutateFunc)
```

Comment:
If there's already an `ExternalSecret` in the target namespace it will be overridden. I think we should reflect such collisions in the status field.

Reply:
How would we check that it's not ours? I know we set controller reference but I'm not seeing a simple way of figuring out if it's something we created or if it existed before.

Reply:
`SetControllerReference()` returns an `AlreadyOwnedError` if someone else claimed ownership of it. If no one owns it then we shouldn't touch it because it has been created by a user.

We should check that if `metadata.ownerReferences` is empty then we won't touch it (and tell the user that this is an issue). I don't think we have an issue with a race condition because we already create the object with a particular `ownerRef`.

We could also set well-known labels or annotations to identify whether or not this is managed by the controller - but i don't think it's strictly necessary.

Reply:
Hey @ADustyOldMuffin, i think this is the only thing that needs to be addressed before we can merge this in. Do you have time to work on that? If not i have some time next week to take this over.

Reply:
So I pushed handling existing secrets, either owned by someone, or not owned by anyone. 

I also added the updating of the status saying wether the ClusterExternalSecret has been successful or not, and if not a list of namespaces that failed.